addons/product/models/product_template.py

```diff
@@ -967,24 +1023,24 @@ def _get_possible_combinations(self, parent_combination=None, necessary_values=N
 
         necessary_values = necessary_values or self.env['product.template.attribute.value']
         necessary_attributes = necessary_values.mapped('attribute_id')
-        ptal_stack = [self.valid_product_template_attribute_line_ids.filtered(lambda ptal: ptal.attribute_id not in necessary_attributes)]
-        combination_stack = [necessary_values]
-
-        # keep going while we have attribute lines to test
-        while len(ptal_stack):
-            attribute_lines = ptal_stack.pop()
-            combination = combination_stack.pop()
-
-            if not attribute_lines:
-                # full combination, if it's possible return it, otherwise skip it
-                if self._is_combination_possible(combination, parent_combination):
-                    yield(combination)
-            else:
-                # we have remaining attribute lines to consider
-                for ptav in reversed(attribute_lines[0].product_template_value_ids):
-                    ptal_stack.append(attribute_lines[1:])
-                    combination_stack.append(combination + ptav)
 
+        attributes = self.valid_product_template_attribute_line_ids.filtered(lambda ptal: ptal.attribute_id not in necessary_attributes)
+
+        if not attributes and self._is_combination_possible(necessary_values, parent_combination):
+            yield necessary_values
+
+        # compute all possible values for each attribute
+        attributes_values = [
+            attribute.product_template_value_ids
+            for attribute in attributes
+        ]
+
+        for product in self._cartesian_product(attributes_values):
```

Comment:
Can we do without this extra method? It's just an indirection that is called only once, I would rather keep all the code related to this in here.

Reply:
I preferred it in a separate method because it's a very specific concern : performing a Cartesian product. You could literally replace that call with `itertools.product` and it would work, albeit more slowly.

Reply:
Well it would not be the same result since yours specifically skips over certain combinations, it's not a full catersian product.

Reply:
@seb-odoo Yes, but you could use `itertools.product(attributes_values)` instead, you'd only loose performance. I can copy paste the code of the new method inside the existing method, but I like the extra method. The extra method handles the performance issue.

Reply:
Now that you pass also parent exclusions, it's not even interchangeable with `itertools.product` by the way.

Reply:
The function is not interchangeable but the expressions `itertools.product(attributes_values)` and `self._cartesian_product(product_template_attribute_values, parent_combination)` are, showing they share a similar concern.

We made the algorithm specific for the module but the concept of a Cartesian product and rejection are not. Here is a generic version of the function :

```python
def my_product(pools: List[List[Any]], rules: Dict[Any, List[Any]], forbidden: List[Any]) -> Iterator[Tuple[Any]]:
    """
    Compute the cartesian product except it rejects some combinations based on provided rules

    :param pools: the elements to calculate the Cartesian product on
    :param rules: a dict specifying which elements each element is incompatible with (key is incompatible with values)
    :param forbidden: elements that are never authorized in the combinations
    :return: the cartesian product without rejected combinations
    """
    if not pools:
        return

    included = set()

    # if an element has an entry of 0, it's acceptable, if greater than 0, it's rejected, cannot be negative
    incompatibles = defaultdict(int)
    for element in forbidden:
        incompatibles[element] += 1
    selections = [-1] * len(pools)
    pool_idx = 0

    def current_element():
        return pools[pool_idx][selections[pool_idx]]

    while True:
        # Discard incompatibilities from element from previous iteration on same pool
        if selections[pool_idx] >= 0:
            for element in rules[current_element()]:
                incompatibles[element] -= 1
            included.discard(current_element())

        # Try to get to next element of same pool
        if selections[pool_idx] != len(pools[pool_idx]) - 1:
            selections[pool_idx] += 1
        # Go to previous pool if current is exhausted
        elif pool_idx != 0:
            selections[pool_idx] = - 1
            pool_idx -= 1
            continue
        # Done if first pool is exhausted
        else:
            break

        # Add incompatibilities of newly added element
        for element in rules[current_element()]:
            incompatibles[element] += 1
        included.add(current_element())

        # Skip element if incompatible
        if incompatibles[current_element()] or \
                any(intersection in included for intersection in rules[current_element()]):
            continue

        # Submit combination if we're at last pool
        if pools[pool_idx] == pools[-1]:
            yield tuple(pool[selection] for pool, selection in zip(pools, selections))
        # Else get to next pool
        else:
            pool_idx += 1
```